### DISCUSSION TITLE
DBeaver: How to get tables metadata

### DISCUSSION THREAD

**Discussion by mmuru** [apache/kyuubi#1189] [Q&A] [answered]
I am using Kyuubi 1.3.0 release with DBeaver 21.1.5. In DBeaver, I could see tables listed in spark_catalog.default schema but it does not show tables metadata (e.g columns and datatypes) and empty. However, I could run `describe table table-name` and returns columns/datatypes. Please, let me know if I missed any steps or configurations? Thanks.

**Comment by yaooqinn** [comment]
can you provide the server and engine logs?

**Reply by mmuru** [reply]
@yaooqinn: Sure, attached log file
[kyuubi-muru-org.apache.kyuubi.server.KyuubiServer-Murus-MacBook-Pro.local.out.zip](https://github.com/apache/incubator-kyuubi/files/7313200/kyuubi-muru-org.apache.kyuubi.server.KyuubiServer-Murus-MacBook-Pro.local.out.zip)

**Comment by pan3793** [comment]
It should be fixed in #1133, would you please try the latest master/branch-1.3?

**Reply by mmuru** [reply]
@pan3793: I tried the latest master and still the same result. It didn't show columns.

**Reply by pan3793** [reply]
Both DBeaver bundled Hive JDBC Driver and Kyuubi provided Hive JDBC driver work well on my laptop.

The difference between these two is that vanilla Hive JDBC driver doesn't support multi catalogs, Kyuubi provided one fixed this issue, you are not in such a case, I think both two should work well.

Using DBeaver bundled Hive JDBC Driver to connect Kyuubi Server

![image](https://user-images.githubusercontent.com/26535726/136721727-e455d562-ce77-44e3-9176-342e92d811a0.png)

![image](https://user-images.githubusercontent.com/26535726/136721980-d458fc46-43b1-4d23-b75a-5e6c1b7ee426.png)

To use Kyuubi provided Hive JDBC driver, please first checkout code from the master branch, then run `build/mvn clean install` to build the latest Kyuubi JDBC jars.

![image](https://user-images.githubusercontent.com/26535726/136491621-ef943d59-a9d7-4e32-ab6b-d5564fd66965.png)

![image](https://user-images.githubusercontent.com/26535726/136489400-457ac0f5-5bd0-4311-bb53-e9168b825c92.png)

**Reply by mmuru** [reply]
@pan3793: I tried Kyuubi hive JDBC driver as per your instructions but still the same issue. It still did not show the columns. I used spark engine version 3.0.3 (as spark-provided). Please, can you provide the engine version information?  

Attached both Kyuubi and spark-engine log files
[kyuubi-spark-sql-engine.log.0.zip](https://github.com/apache/incubator-kyuubi/files/7319429/kyuubi-spark-sql-engine.log.0.zip)
[kyuubi-muru-org.apache.kyuubi.server.KyuubiServer-Murus-MacBook-Pro.local.out.zip](https://github.com/apache/incubator-kyuubi/files/7319431/kyuubi-muru-org.apache.kyuubi.server.KyuubiServer-Murus-MacBook-Pro.local.out.zip)

.

**Reply by pan3793** [reply]
Tested with Spark 3.1.2, let me try 3.0.3.

**Reply by mmuru** [reply]
FYI: I tried Spark 3.1.2 too but the result was the same. Using the Spark bundled Hadoop 3.2.0 and Hive 2.3.7. I am using Postgresql as Hive metastore. Please, let me know if you need more information.

**Comment by yaooqinn** [comment]
![image](https://user-images.githubusercontent.com/8326978/136667797-c4832fd8-edbc-4b45-9f6f-dde80e93ab71.png)

It works for me on my laptop

**Reply by mmuru** [reply]
@yaooqinn & @pan3793 

I am using Delta-lake table.  I tried Kyuubi 1.3.1, Spark 3.1.2, Deltalake 1.0.0, Hive 2.3.7 (HiveMetastoreConnection version 2.3.7 using Spark classes). Please, let me know if you need more information.
````
CREATE TABLE conn_stats (
month STRING, 
item_count_before bigint,
item_count_after bigint
)
USING DELTA OPTIONS (path 's3a://delta/connection-summary/');
````
spark.catalog.listTables()
````
Table(name='conn_stats', database='default', description=None, tableType='EXTERNAL', isTemporary=False)
````
In DBeaver: describe table spark_catalog.default.conn_stats; 
````
+--------------------+---------+-------+                                        
|            col_name|data_type|comment|
+--------------------+---------+-------+
|               month|   string|       |
|item_count_before   |   bigint|       |
|item_count_after    |   bigint|       |
|                    |         |       |
|      # Partitioning|         |       |
|     Not partitioned|         |       |
+--------------------+---------+-------+
````

**Reply by pan3793** [reply]
Previous answers based on Hive tables and Iceberg tables, let me try Delta tables

**Reply by pan3793** [reply]
@mmuru I can reproduce with delta table, will investigate it later

**Reply by mmuru** [reply]
@pan3793: Thanks for the update and looking into it. Please, let me know if you want me to create issue for this.

**Reply by yaooqinn** [reply]
Hi @mmuru, sounds good to me to create an issue to track this.

> @pan3793: Thanks for the update and looking into it. Please, let me know if you want me to create issue for this.

**Reply by pan3793** [reply]
After some investigation, I found it's a known issue of delta https://github.com/delta-io/delta/issues/573

**Comment by pan3793** [comment, marked as answer]
Based on delta v1.0.0 source code, delta will erase schema before persisting table information into Hive metastore. The behavior doesn't change in master branch.

https://github.com/delta-io/delta/blob/5571d95966e3b03c4699dcfa1a5cd86bb16447dc/core/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala#L387-L405

DBevaer uses `GET_COLUMNS` operation to ask Kyuubi for table column information, and Kyuubi handles it by calling Spark Catalog API, which retrieves metadata from Hive metastore, then you got nothing.

**Reply by mmuru** [reply]
@pan3793: Thanks for the update. Please, can you point me to where in Kyuubi side calling Spark catalog API? Is it something Kyuubi
side can handle it?

**Reply by pan3793** [reply]
> where in Kyuubi side calling Spark catalog API?

https://github.com/apache/incubator-kyuubi/blob/ddfd0355e8867c7ea51f511a0c0023c3bf0e7e6e/externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/shim/CatalogShim_v2_4.scala#L91-L104

> Is it something Kyuubi side can handle it?

Technically, Kyuubi can, but not reasonable, it's a delta issue.

Kyuubi can use `spark.table(xxx).schema` to retrieve table schema to suppress this issue, but it makes `GET_COLUMNS` be a heavy operation rather than a lightweight metadata operation.

**Reply by mmuru** [reply]
@pan3793: Thanks for the details. I understand it is delta side issue. In the meantime, we could fix it in Kyuubi side till it is addressed in delta side.
